backend/src/index.ts

```diff
@@ -1,54 +1,94 @@
-import express from 'express';
-
-import { collection, addDoc, getDocs } from 'firebase/firestore';
-import { db } from './firebase';
-import cors from 'cors';
+import express, { Request, Response } from "express";
+import { TypedRequest, LoginBody } from "./requestTypes";
+import bcrypt from "bcrypt";
+import { collection, addDoc, getDocs, where, query } from "firebase/firestore";
+import { db } from "./firebase";
+import cors from "cors";
 
 const EXPRESS_PORT = 3000;
 
 const app = express();
 app.use(express.urlencoded({ extended: true }));
-app.use(express.json())
+app.use(express.json());
 app.use(cors());
 
-app.get('/', (req, res) => {
-  res.send('yellowshirt backend says hello');
+app.get("/", (req, res) => {
+  res.send("yellowshirt backend says hello");
+});
+
+app.post("/register", async (req: TypedRequest<LoginBody>, res: Response) => {
+  const { username, password } = req.body;
+
+  const querySnapshot = await getDocs(collection(db, "users"));
+  if (querySnapshot.docs.some((doc) => doc.data().username === username)) {
+    return res.status(400).send("Username already exists");
+  }
+
+  const saltRounds: number = 10;
+  const hashedPassword: string = await bcrypt.hash(password, saltRounds);
+  await addDoc(collection(db, "users"), {
+    username,
+    password: hashedPassword,
+  });
+
+  res.status(201).send("User Successfully Registered");
+});
+
+app.post("/login", async (req: TypedRequest<LoginBody>, res: Response) => {
+  const { username, password } = req.body;
+  const users = collection(db, "users");
+  const loginDetails = query(users, where("username", "==", username));
+  const details = await getDocs(loginDetails);
+
+  if (details.empty) {
+    return res.status(400).send("Username not found");
+  }
+
+  bcrypt.compare(password, details.docs[0].data().password, (err, result) => {
```

Comment:
just note that if we add a salt above, we'll also need to add it here to the entered pw